models/darknet/darknet.hpp

```diff
@@ -0,0 +1,305 @@
+   * @param strideWidth Stride of filter application in the x direction.
+   * @param strideHeight Stride of filter application in the y direction.
+   * @param padW Padding width of the input.
+   * @param padH Padding height of the input.
+   */
+  Sequential<>* ConvolutionBlock(const size_t inSize,
+                                 const size_t outSize,
+                                 const size_t kernelWidth,
+                                 const size_t kernelHeight,
+                                 const size_t strideWidth = 1,
+                                 const size_t strideHeight = 1,
+                                 const size_t padW = 0,
+                                 const size_t padH = 0,
+                                 const size_t batchNorm = false)
+  {
+    Sequential<>* bottleNeck = new Sequential<>();
+    bottleNeck->Add(new Convolution<>(inSize, outSize, kernelWidth,
+        kernelHeight, strideWidth, strideHeight, padW, padH, inputWidth,
+        inputHeight));
+    if (batchNorm)
+    {
+      bottleNeck->Add(new BatchNorm<>(outSize));
+    }
+
+    bottleNeck->Add(new LeakyReLU<>());
+
+    // Update inputWidth and input Height.
+    std::cout << "Conv Layer.  ";
+    std::cout << "(" << inputWidth << ", " << inputHeight <<
+        ", " << inSize << ") ----> ";
+
+    inputWidth = ConvOutSize(inputWidth, kernelWidth, strideWidth, padW);
+    inputHeight = ConvOutSize(inputHeight, kernelHeight, strideHeight, padH);
+    std::cout << "(" << inputWidth << ", " << inputHeight <<
+        ", " << outSize << ")" << std::endl;
+    return bottleNeck;
+  }
+
+  /**
+   * Adds Pooling Block.
+   *
+   * @param kernelWidth Width of the filter/kernel.
+   * @param kernelHeight Height of the filter/kernel.
+   * @param strideWidth Stride of filter application in the x direction.
+   * @param strideHeight Stride of filter application in the y direction.
+   */
+  Sequential<>* PoolingBlock(const size_t kernelWidth,
+                             const size_t kernelHeight,
+                             const size_t strideWidth = 1,
+                             const size_t strideHeight = 1,
+                             const std::string type = "max")
+  {
+    Sequential<>* bottleNeck = new Sequential<>();
+    if (type == "max")
+    {
+      bottleNeck->Add(new MaxPooling<>(kernelWidth, kernelHeight,
+        strideWidth, strideHeight, true));
+    }
+    else
+    {
+      bottleNeck->Add(new MeanPooling<>(kernelWidth, kernelHeight,
+        strideWidth, strideHeight, true));
+    }
+    std::cout << "Pooling Layer.  ";
+    std::cout << "(" << inputWidth << ", " << inputHeight <<
+        ") ----> ";
+    // Update inputWidth and inputHeight.
+    inputWidth = PoolOutSize(inputWidth, kernelWidth, strideWidth);
+    inputHeight = PoolOutSize(inputHeight, kernelHeight, strideHeight);
+    std::cout << "(" << inputWidth << ", " << inputHeight <<
+        ")" << std::endl;;
+    return bottleNeck;
```

Comment:
@KimSangYeon-DGU, is this not a valid operation?

Reply:
Since I didn't wan't to make changes in this branch I was using a different branch. I just pushed it [here](https://github.com/kartikdutt18/models/tree/BreakingDownDarknet)

Reply:
It seems I successfully re-produced the issue. I'll look into it.
<img width="460" alt="image" src="https://user-images.githubusercontent.com/26578203/84562251-de0d2080-ad07-11ea-9507-79154548442f.png">

Reply:
Great, Thanks a lot.

Reply:
VGDB didn't get me anywhere, maybe instead of returning the pointer I should add it to the base model. Maybe that should work. I'll try that tomorrow morning.

Reply:
> maybe instead of returning the pointer I should add it to the base model. 

Yes, I think so. let's try do that.

Reply:
@KimSangYeon-DGU, To resolve the segmentation fault, I stopped returning the pointer and avoided copying FFN. However it gives me this error now,
```
Assertion failed: (false), function forced_return, file /usr/local/include/boost/variant/detail/forced_return.hpp, line 38.
```

Reply:
Hmm, So the problem occurs in adding a pooling layers. If a make a model of DarknetBlocks and Convolutional blocks it works. Somehow pooling layers cause the error above.

Reply:
@kartikdutt18 I was trying to this code, but it failed to build because of the `no matching function for call to 'mlpack::ann::Convolution<...`. It's already 4:40 AM to me, so I'll look into it further in the night

```
if (DarkNetVer == 19)
  {
    darkNet.Add<IdentityLayer<> >();

    darkNet.Add<Convolution<> >(inputChannel, 32, 3, 3, 1, 1, 1, 1, 128, 128);
    darkNet.Add<LeakyReLU<> >();

    darkNet.Add<MaxPooling<> >(2, 2, 2, 2, true);

    darkNet.Add<Convolution<> >(32, 64, 3, 3, 1, 1, 1, 1, 64, 64);
    darkNet.Add<LeakyReLU<> >();

    darkNet.Add<MaxPooling<> >(2, 2, 2, 2, true);

    darkNet.Add<Convolution<> >(64, 128, 3, 3, 1, 1, 1, 1, 1, 32, 32);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(128, 64, 1, 1, 1, 1, 1, 1, 1, 32, 32);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(64, 128, 3, 3, 1, 1, 1, 1, 1, 32, 32);
    darkNet.Add<LeakyReLU<> >();

    darkNet.Add<MaxPooling<> >(2, 2, 2, 2, true);

    darkNet.Add<Convolution<> >(128, 256, 3, 3, 1, 1, 1, 1, 1, 16, 16);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(256, 128, 1, 1, 1, 1, 1, 1, 1, 16, 16);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(128, 256, 3, 3, 1, 1, 1, 1, 1, 16, 16);
    darkNet.Add<LeakyReLU<> >();

    darkNet.Add<MaxPooling<> >(2, 2, 2, 2, true);

    darkNet.Add<Convolution<> >(256, 512, 3, 3, 1, 1, 1, 1, 1, 8, 8);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(512, 256, 1, 1, 1, 1, 1, 1, 1, 8, 8);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(256, 512, 3, 3, 1, 1, 1, 1, 1, 8, 8);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(512, 256, 1, 1, 1, 1, 1, 1, 1, 8, 8);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(256, 512, 3, 3, 1, 1, 1, 1, 1, 8, 8);
    darkNet.Add<LeakyReLU<> >();

    darkNet.Add<MaxPooling<> >(2, 2, 2, 2, true);

    darkNet.Add<Convolution<> >(512, 1024, 3, 3, 1, 1, 1, 1, 1, 4, 4);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(1024, 512, 1, 1, 1, 1, 1, 1, 1, 4, 4);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(512, 1024, 3, 3, 1, 1, 1, 1, 1, 4, 4);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(1024, 512, 1, 1, 1, 1, 1, 1, 1, 4, 4);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(512, 1024, 3, 3, 1, 1, 1, 1, 1, 4, 4);
    darkNet.Add<LeakyReLU<> >();
    darkNet.Add<Convolution<> >(1024, 1000, 1, 1, 1, 1, 1, 1, 1, 4, 4);
    darkNet.Add<LeakyReLU<> >();

    //darkNet.Add(PoolingBlock(inputWidth, inputHeight,
    //    inputWidth, inputHeight, "mean"));

    if (includeTop)
    {
      darkNet.Add<Linear<> >(1000, numClasses);
      darkNet.Add<LogSoftMax<> >();
    }

    darkNet.ResetParameters();
  }
```

Reply:
Also, I found the initialization with template paramters `FFN<OutputLayerType, InitializationRuleType> darkNet;` doesn't work properly, so I manually set it like below for the debugging.
```
FFN<NegativeLogLikelihood<>, RandomInitialization> darkNet;
```

Reply:
> Hmm, So the problem occurs in adding a pooling layers. If a make a model of DarknetBlocks and Convolutional blocks it works. Somehow pooling layers cause the error above.

Ahh, Ok, that's good finding!

Reply:
I think you are on Linux machine, So in Linux machine the compiler gets confused between Add layer and Add function. So some this like would work : 
```
if (DarkNetVer == 19)
  {
    darkNet.Add(new IdentityLayer<>());

    darkNet.Add(new Convolution<>(inputChannel, 32, 3, 3, 1, 1, 1, 1, 128, 128));
    darkNet.Add(new LeakyReLU<>());

    darkNet.Add(new MaxPooling<>(2, 2, 2, 2, true));
```

Reply:
> Also, I found the initialization with template paramters FFN<OutputLayerType, InitializationRuleType> darkNet; doesn't work properly, so I manually set it like below for the debugging.

That's strange. Right now I just compiled mlpack and added input variable (width / height) in constructor, incase the issue was with some visitor but even then the code fails with the same error.

Reply:
I think I got the model working on rough implementation, I'll just clean it up and start training.

Reply:
On 128 x128, These are the results of 3 epochs. A smaller size like 56 or something would give better results.
```
5.08502e+9================================================================================================] 100% - ETA: 0s - loss: 4.89669e+9
26/26 [====================================================================================================] 100% - 49s 6529ms/step - loss: 5.08502e+9
Epoch 2/3
1.65297e+9================================================================================================] 100% - ETA: 0s - loss: 1.59175e+9
26/26 [====================================================================================================] 100% - 52s 6636ms/step - loss: 1.65297e+9
Epoch 3/3
6.2459e+8=================================================================================================] 100% - ETA: 0s - loss: 6.01457e+8
26/26 [====================================================================================================] 100% - 50s 6563ms/step - loss: 6.2459e+8
```

Reply:
Will clean up, push a commit and start training.

Reply:
So we can't use 32 x 32 image,
```
Conv Layer.  (32, 32, 3) ----> (32, 32, 32)
Conv Layer.  (16, 16, 32) ----> (16, 16, 64)
Conv Layer.  (8, 8, 64) ----> (8, 8, 128)
Conv Layer.  (8, 8, 128) ----> (8, 8, 64)
Conv Layer.  (8, 8, 64) ----> (8, 8, 128)
Conv Layer.  (2, 2, 128) ----> (2, 2, 256)
Conv Layer.  (2, 2, 256) ----> (2, 2, 128)
Conv Layer.  (2, 2, 128) ----> (2, 2, 256)
Conv Layer.  (1, 1, 256) ----> (1, 1, 512)
Conv Layer.  (1, 1, 512) ----> (1, 1, 256)
Conv Layer.  (1, 1, 256) ----> (1, 1, 512)
Conv Layer.  (1, 1, 512) ----> (1, 1, 256)
Conv Layer.  (1, 1, 256) ----> (1, 1, 512)
Conv Layer.  (0, 0, 512) ----> (0, 0, 1024)
Conv Layer.  (0, 0, 1024) ----> (0, 0, 512)
Conv Layer.  (0, 0, 512) ----> (0, 0, 1024)
Conv Layer.  (0, 0, 1024) ----> (0, 0, 512)
Conv Layer.  (0, 0, 512) ----> (0, 0, 1024)
Conv Layer.  (0, 0, 1024) ----> (0, 0, 1000)
```

Reply:
Hmm, can you try to use padding? Not sure if this is a good solution though

Reply:
Agreed, I can try using valid padding type in pooling layers. Let me use that. I also tried 48 and 56 they also didn't work.

Reply:
So I used adaptive pooling instread and set input dim => ceil(input dim / 2) instead which makes it work for 32 x 32.
```
Conv Layer.  (32, 32, 3) ----> (32, 32, 32)
Pooling Layer.  (32, 32) ----> (16, 16)
Conv Layer.  (16, 16, 32) ----> (16, 16, 64)
Pooling Layer.  (16, 16) ----> (8, 8)
Conv Layer.  (8, 8, 64) ----> (8, 8, 128)
Conv Layer.  (8, 8, 128) ----> (8, 8, 64)
Conv Layer.  (8, 8, 64) ----> (8, 8, 128)
Pooling Layer.  (8, 8) ----> (4, 4)
Conv Layer.  (4, 4, 128) ----> (4, 4, 256)
Conv Layer.  (4, 4, 256) ----> (4, 4, 128)
Conv Layer.  (4, 4, 128) ----> (4, 4, 256)
Pooling Layer.  (4, 4) ----> (2, 2)
Conv Layer.  (2, 2, 256) ----> (2, 2, 512)
Conv Layer.  (2, 2, 512) ----> (2, 2, 256)
Conv Layer.  (2, 2, 256) ----> (2, 2, 512)
Conv Layer.  (2, 2, 512) ----> (2, 2, 256)
Conv Layer.  (2, 2, 256) ----> (2, 2, 512)
Pooling Layer.  (2, 2) ----> (1, 1)
Conv Layer.  (1, 1, 512) ----> (1, 1, 1024)
Conv Layer.  (1, 1, 1024) ----> (1, 1, 512)
Conv Layer.  (1, 1, 512) ----> (1, 1, 1024)
Conv Layer.  (1, 1, 1024) ----> (1, 1, 512)
Conv Layer.  (1, 1, 512) ----> (1, 1, 1024)
Conv Layer.  (1, 1, 1024) ----> (1, 1, 1000)
```

Reply:
These results are on subset (1k images -> 32 x 32), Kindly let me know what you think,
```
Epoch 1/3
2.28237====================================================================================================] 100% - ETA: 0s - loss: 2.19784
26/26 [====================================================================================================] 100% - 31s 5844ms/step - loss: 2.28237
Epoch 2/3
2.23357====================================================================================================] 100% - ETA: 0s - loss: 2.15084
26/26 [====================================================================================================] 100% - 30s 5795ms/step - loss: 2.23357
Epoch 3/3
2.18636====================================================================================================] 100% - ETA: 0s - loss: 2.10539
26/26 [====================================================================================================] 100% - 30s 5773ms/step - loss: 2.18636
```
I also switched to He Initialization from Random.

Reply:
I guess the loss is decreasing which is good :) Can you also report the classification accuracy?

Reply:
Sure, I'll write a custom callback for it.

Reply:
I have implemented the callback, The output looks something like this, (3.14 is just a dummy value). Pushing a commit for it.
```
Epoch 1/3
2.28237====================================================================================================] 100% - ETA: 0s - loss: 2.19784
26/26 [====================================================================================================] 100% - 31s 5812ms/step - loss: 2.28237
train accuracy : 3.140000
```

Reply:
I totally agree to check the progress of the training by printing the classification accuracy against the validation dataset, but we should consider whether this feature is vital or not at this juncture, because we're on the beginning of the summer and there might be critical blocker in the middle of working.

At this point, I think checking the accuracy against the test dataset using the trained model is enough.

IMO, for the time being, we can open a PR about that, and then we can work this later after finishing other works scheduled.

Reply:
> IMO, for the time being, we can open a PR about that, and then we can work this later after finishing other works scheduled.

Makes sense. So I'll keep it as ensmallen_utils.hpp for now and open a PR in ensmallen that can be completed when other tasks are completed.

Reply:
And I think it's very nice to check why `MaxPooling` and `MeanPooling` don't work, but `AdaptiveMaxPooling` and `AdaptiveMeanPooling` works.

Reply:
Ahh, sorry @kartikdutt18 I wanted to mean `issue`, not PR.

Reply:
I got MaxPooling and MeanPooling to work as well but for 32 x 32 they go to 0 x 0 so I used adaptive pooling to maintain a size of atleast 1.

Reply:
Ok, thanks for correcting me. Can you open a issue about the ensmallen's callback? I think we don't need to open a PR at this time to make a space other contributors take

Reply:
Sure, I can open an issue.

Reply:
Opened mlpack/ensmallen#200

Reply:
Initial result on subset.
```
Epoch 1/3
2.28237====================================================================================================] 100% - ETA: 0s - loss: 2.19784
[====================================================================================================] 100% - 36s 6013ms/step - loss: 2.28237
Train accuracy : 30.769231
```
I guess I can start training on whole dataset now. Should also print validation accuracy.

Reply:
Alright great! 👍  How long is the training going to take? I thought it would be quite slow since the model is quite big.

Reply:
It takes about 10 minutes to load the dataset. Once 2 - 3 epochs complete, I will post the results and a rough idea about training time. I will just remove all redundant files from my system first.

Reply:
alright sure. Sounds good.

Reply:
Started training, will post the results soon.

Reply:
It's been roughly an hour now since model started training. An it is still at the first epoch.
```
68/998 [======>.............................................................................................] 6% - ETA: 48s - loss: 2.15807
```

Reply:
why is the ETA so off?

Reply:
I'm not sure, that is the ETA for the iteration or the epoch?
I think it is for a single iteration.

Reply:
alright makes sense

Reply:
Great, I'll let it train the whole night. I'll also plot a graph for the results tomorrow.

Reply:
I think the model is overfitting : 
Train accuracy : 75.88
One epoch took nearly 12 epochs or more. Should I train on the subset first?

Reply:
I am not sure if that would work. If the model is overfitting then the train accuracy should have been higher right?
I think we could try training for one more epoch to see if there is any improvement in the results. Let me know what you think.

Reply:
> If the model is overfitting then the train accuracy should have been higher right?
I think that model overfit because even simple models like LeNet on MNIST also don't reach 75% accuracy.

Sure, I have added a periodic save so that we at least have weights after each epoch. So I have set the epochs = 5 for now, earlier I had set them equal to 10 and we wouldn't have gotten any weights until it finished training. Will keep posting results here.

Reply:
Also added validation accuracy after each epoch so we will know if the model overfits, if the difference is too large to ignore.

Reply:
Great! thanks for the update. I also agree with the idea of training on the subset.

Reply:
Great, The current size is 100 images per class so 1000 images. Should I go for 10k or 15k?
Kindly let me know what makes sense.

Reply:
You'll increase the size of images per class to make the entire training size 10k or 15k, right? Then, let's go for 15k :)

Reply:
Great, So should I stop the current training on whole dataset?

Reply:
And I have a question. How will you make the subset?

Reply:
I'll select the first 1500 images (in sorted form) from training set of cifar10 training set.

Reply:
BTW, is it possible to increase the size of the training dataset? To prevent the model from overfitting, it's better to train the model with images as many as possible.

Reply:
> Great, The current size is 100 images per class so 1000 images. Should I go for 10k or 15k?

It seems you already made a subset :)

Reply:
With 40k images for training it takes more than 12 hours for a single epoch. I can let it train on any number between 10k and 40k. What do you suggest.

Reply:
This 100 images per class or 1k subset wad just to test callbacks and model.

Reply:
> > Great, The current size is 100 images per class so 1000 images. Should I go for 10k or 15k?
> 
> It seems you already made a subset from the 50,000 training dataset :)

Yup that's what I was thinking. 
What was the final train accuracy you got?

Reply:
> This 100 images per class or 1k subset was just to test callbacks and model.

So, when you trained the model with 75.88 accuracy, how many training and testing dataset did you use?

Reply:
Its still training on 40k images. I had to add periodic save so I had to restart training otherwise we wouldn't have gotten any weights for the next 10 days as well valid accuracy. Its half way through the first epoch. We should have training accuracy for 2 epochs by tomorrow afternoon.

Reply:
For actual training I used the orignal dataset with 80% training data(40k) and 20% valid data(10k).

Reply:
@saksham189 @kartikdutt18 How about train with 10k images, not 40k? 40k take so long time...

Reply:
Makes sense. That way we could have multiple epochs in a day.

Reply:
> Makes sense. That way we could have multiple epochs in a day.

Let's give it a try :)

Reply:
Sure I will make the changes.

Reply:
I think it willl be even faster now, looks like OpenMP wasn't being linked correctly in cmake.
Fixed using:
```
if (USE_OPENMP AND OPENMP_FOUND AND CMAKE_CXX_COMPILER_ID MATCHES "AppleClang")
  if (CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "7.0" OR
          CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "7.0")
        target_link_libraries(${name} ${OpenMP_libomp_LIBRARY})
      endif()
  endif()
```